### PR TITLE
refactor: decouple core from Tauri, extract note session, split Rust shell

### PR DESCRIPTION
**Pull request by maccman**
## What

A structural refactor converging the codebase with its own stated architecture (`docs/plans/architecture-conventions.md`), aimed at the "high-quality open source" bar: decoupled components, documented APIs, tests at the seams. Net **−1,300 lines** despite adding ~70 tests and four docs. Includes a merge of master (Plan 06) with conflicts resolved into the new structure.

### 1. Core decoupled from Tauri
- New `IpcBridge` seam (`packages/core/src/ipc/bridge.ts`): `@reflect/core` and `@reflect/db` no longer depend on `@tauri-apps/*` at all. The desktop app installs the one Tauri adapter at startup (`apps/desktop/src/lib/tauri-bridge.ts`); tests install in-memory fakes via `setBridge()` instead of `vi.mock`.
- `@reflect/db`'s dialect takes an injected `QueryRunner`; core owns the shared Kysely instance wired to the bridge.
- `watch.ts` split into `file-changes.ts` (subscription + zod validation) and `live.ts` (re-index orchestration, injectable error handler).
- This is the prerequisite for the Plan 14 CLI ("reuses @reflect/core without the Rust process").

### 2. NoteSession state machine
- The 404-line `use-note-document.ts` hook's save/conflict/protection pipeline is now a pure, React-free state machine (`editor/note-session.ts`) with injected IO. One session per path eliminates the cross-note ref guards — the path-switch "final flush" is just `dispose()`.
- The hook is a ~140-line adapter; all pre-existing hook tests pass unchanged.
- Image persistence (`use-image-persistence.ts`) and app-version fetch extracted from components; failed image saves now surface in the UI instead of console-only.

### 3. Rust shell split + hardening
- `db.rs` (622 lines) → `db/{mod,migrations,query,write,tests}.rs`; `fs.rs` (471) → `fs/{mod,resolve,io}.rs`, tests beside their modules.
- `eprintln!` → `tracing` (honors `RUST_LOG`); lock-poison errors log context.
- Watcher race closed at the source: `watch_start` holds the graph lock from root-read through debouncer install (chosen over generation-tagging events, which would entangle the graph and index generation counters across the IPC contract).

### 4. Tests at the seams (~70 new)
- Core: `toAppError` edge cases, the `call()` validation choke point, watcher batch serialization/malformed payloads, the wiki-link Lezer grammar directly.
- Desktop: NoteSession contracts (dispose semantics, snapshot dedupe, keepMine-on-identical-content), `GraphProvider` open-sequencing (overlapping opens serialize, last wins), GraphChooser render flows.
- Rust: command-level test via `tauri::test` proving stale-generation writes are dropped end-to-end.

### 5. OSS docs + cleanups
- Real root README (was the Tauri template stub), `packages/core` + `packages/db` READMEs, `CONTRIBUTING.md`, a "Plan NN" legend, stable-vs-plumbing annotations in the core barrel.
- Duplicated link regex unified in `markdown/link-syntax.ts` (shared by indexer + editor scanners).
- TanStack Query adoption recorded in Plan 07 (deferred from this refactor by design).

### Merge of master (Plan 06)
Plan 06 landed on the same files mid-refactor; conflicts were resolved by porting its behavior into the new structure:
- `createIfMissing` (lazy daily notes) lives in the NoteSession machine, applied to the **initial load only** — a delete racing reconcile still no-ops rather than emptying the editor. Master's lazy-note tests pass unchanged.
- `note-pane` keeps Plan 06's lazy/autoFocus/wiki-link navigation on top of the extracted image hook; `graph-workspace` takes the router-driven layout with the extracted hooks. `use-initial-note-path` was deleted — the daily stream replaces the welcome-note bootstrap.

## Reviewer notes

- **No behavior changes intended** outside: (a) image-save failures now visible in the pane, (b) lock-poison/asset-scope logging goes through `tracing`, (c) the watcher install race fix.
- The `IndexedNote` field visibility in Rust went from private to `pub(super)` solely so the relocated test module can construct fixtures.
- Deliberately skipped: retryable-vs-fatal error kinds (keeps the TS↔Rust `AppError` contract stable) and a big-bang `actions/<domain>/` restructure (the conventions doc says adopt per-domain as Plans 06+ land).

## Verification

- `pnpm typecheck`, `oxlint`: clean
- TS: 188 tests passing (96 core, 4 db, 88 desktop — includes Plan 06's new tests)
- Rust: 26 tests passing, clippy clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches graph/index generation gating, file watcher install ordering, and the note save pipeline—core data integrity paths—but behavior is largely preserved with new tests at the boundaries.
> 
> **Overview**
> **Decouples `@reflect/core` from Tauri** via an injected `IpcBridge` (`setBridge` / `hasBridge`): the desktop app installs a thin adapter at startup; tests use fakes instead of `vi.mock` on `@tauri-apps/api`. Watcher-driven indexing moves from `watch.ts` into `file-changes.ts` + `live.ts`, with serialized batch application and clearer error handling.
> 
> **Extracts note save/conflict logic** into a pure `createNoteSession` state machine; `useNoteDocument` becomes a thin adapter keyed by path (generation read at write time so same-graph reopen keeps unsaved edits). Image paste/drop moves to `useImagePersistence` with **in-pane alerts** on failed saves and safer `assets/` URL resolution.
> 
> **Splits the Rust shell**: monolithic `db.rs` / `fs.rs` become `db/*` and `fs/*` modules; adds **`tracing`** (replaces `eprintln!`), holds the graph lock through `watch_start` install, and an end-to-end test that stale index generations no-op writes.
> 
> **Docs and tests**: replaces the template README with Reflect/CONTRIBUTING/core READMEs; adds ~70 tests (IPC `call`, graph open sequencing, NoteSession, GraphChooser, Rust generation gate). Unifies inline link parsing in `markdown/link-syntax.ts` for indexer + editor.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b467e08fd8bcf393f1f9050256245d95633cc0f1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Graph/file operations: create/open graphs, read/write/move/delete notes and assets
  * Local SQLite index with read-only query bridge and generation-gated reindexing
  * Note session state machine for safer saves, conflict handling, and background reconciliation
  * Image persistence for pasted/dropped images and related UI error alert
  * Tauri bridge installation and improved runtime logging

* **Documentation**
  * Expanded README and CONTRIBUTING with contribution rules, architecture, and plans

* **Tests**
  * Added comprehensive tests for editor sessions, indexing, FS primitives, and IPC/query plumbing

### CODE REVIEW

File: apps/desktop/src/editor/use-note-document.ts [hunk 1]
```diff
@@ -67,280 +48,43 @@ export function useNoteDocument(
...
-      active = false
+      if (sessionRef.current === session) {
+        sessionRef.current = null
+      }
+      // Disposal flushes pending edits to the session's own path — the
+      // path-switch "final flush" lives here, not in cross-note bookkeeping.
+      session.dispose()
     }
-  }, [path, createIfMissing, reconcileFromDisk])
+  }, [path, generation, createIfMissing])
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts]: Fixed in 0a3b9a5 — real regression, thanks. The write closure now reads the generation from a ref at write time and the session effect no longer keys on `generation`, so a same-graph reopen keeps the session (and dirty buffer) alive and the pending save lands with the fresh generation. Cross-graph pinning is preserved: a graph switch remounts the workspace (keyed by root), so an unmounted pane's final flush still carries its old generation and is rejected. Added a regression test (edit → generation bump → dirty survives, save carries the new generation).

File: apps/desktop/src-tauri/src/db/tests.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    super::index_apply(note("notes/b.md", "B", vec![]), stale, app.state())
+        .expect("stale apply returns Ok");
+    assert_eq!(count("after stale apply"), Value::from(1)); // dropped, not applied
+
+    super::index_remove("notes/a.md".to_string(), stale, app.state())
+        .expect("stale remove returns Ok");
+    assert_eq!(count("after stale remove"), Value::from(1)); // also dropped
+
+    super::index_apply(note("notes/b.md", "B", vec![]), fresh, app.state())
+        .expect("fresh apply");
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/db/tests.rs]: Fixed in 6e4c799 (cargo fmt across the crate).

File: apps/desktop/src-tauri/src/watcher.rs [hunk 3]
```diff
@@ -69,23 +69,30 @@ fn collect_changes(paths: &[PathBuf], root: &Path) -> Vec<FileChange> {
...
     watcher: State<WatcherState>,
 ) -> AppResult<()> {
-    let root = graph
-        .0
-        .lock()
-        .map_err(|_| AppError::io("graph state lock poisoned"))?
+    let graph_guard = graph.0.lock().map_err(|err| {
+        tracing::error!(?err, "graph state lock poisoned by an earlier panic");
+        AppError::io("graph state lock poisoned")
+    })?;
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/watcher.rs:94]: Fixed in 6e4c799 (cargo fmt across the crate).

File: apps/desktop/src/editor/use-image-persistence.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  const [saveError, setSaveError] = useState<string | null>(null)
+
+  const resolveUrl = useCallback(
+    (src: string): string | null => {
+      if (/^https?:\/\//.test(src)) {
+        return src
+      }
+      if (graphRoot && src.startsWith('assets/')) {
+        return convertFileSrc(`${graphRoot}/${src}`)
+      }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-image-persistence.ts]: Fixed in b467e08 — resolveUrl now rejects assets/ sources containing traversal segments (isSafeAssetSource) before calling convertFileSrc. Defense in depth: the Rust shell already canonicalizes and rejects traversal on every write.

File: apps/desktop/src/editor/use-image-persistence.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  )
+
+  const saveImage = useCallback(
+    async (file: File): Promise<string | null> => {
+      const extension = EXTENSION_BY_MIME[file.type]
+      if (!extension || generation === null) {
+        return null
+      }
+      const target = assetPath(`pasted-${Date.now()}.${extension}`)
+      await writeAsset(target, base64Of(await file.arrayBuffer()), generation)
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-image-persistence.ts]: Fixed in b467e08 — pasted-asset names are now `pasted-<timestamp>-<random8>.<ext>`, so same-millisecond pastes (multi-file drops) can't collide.

File: apps/desktop/src/editor/note-session.ts [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    // live editor whose next save would drop what it can't model.
+    adoptCleanContent(content)
+  }
+
+  function dispose(): void {
+    // Flush first: the queued save step reads the (now frozen) buffer, so
+    // pending edits persist to this session's path even after the UI moves on.
+    flush()
+    disposed = true
+  }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/note-session.ts:400]: Not taking this one: React effect cleanup (where dispose runs) cannot await, so an awaitable dispose can't actually sequence teardown against the next session without delaying its mount. The rapid A→B→A race it targets is already non-destructive — the in-flight final write triggers a watcher event, and the new session's reconcile path either reloads cleanly or parks a conflict if the user typed (never silently overwrites). That echo/reconcile loop is the designed convergence mechanism for all out-of-band writes, including our own.

File: apps/desktop/src/editor/use-image-persistence.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ * `generation` so a save racing a graph switch can't land in the wrong graph.
+ * A failed save surfaces on {@link ImagePersistence.saveError} so the pane can
+ * tell the user the image was not stored.
+ */
+export function useImagePersistence(
+  graphRoot: string | null,
+  generation: number | null,
+): ImagePersistence {
+  const [saveError, setSaveError] = useState<string | null>(null)
+
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-image-persistence.ts:66]: Not needed in practice: the workspace tree is keyed by graph root, so switching graphs remounts NotePane (and this hook) with fresh state — saveError can't leak across graphs. Within one graph, the next successful save already clears it.

File: packages/db/src/dialect.ts [hunk 8]
```diff
@@ -11,27 +10,35 @@ import {
...
+    const rows = await this.runQuery(compiled.sql, compiled.parameters)
     // Index reads are our own projection (Rust serializes from a known schema),
     // so per Plan 04 §2 we deliberately don't zod-parse every row (real overhead
     // on large FTS scans). A cheap O(1) shape check still fails fast on a
     // malformed payload at the boundary rather than deep in a query consumer.
     if (!Array.isArray(rows)) {
       throw new Error('db_query did not return a row array')
     }
-    return { rows }
+    return { rows: rows as R[] }
```
[the remark this answers is not in the document]
Comment (maccman) [reply packages/db/src/dialect.ts:41]: Intentionally kept as-is: Plan 04 §2 (and packages/db/README.md) document the decision not to validate our own projection's rows — Rust serializes Vec<Map<String, Value>>, so rows are objects by construction, and the array check catches the realistic failure (wrong payload shape entirely). The `rows as R[]` cast is the same trust boundary the previous `invoke<R[]>` typing expressed; Kysely supplies no runtime schema to validate against without re-stating one per query.
